File: plugins/openwrt/ip_monitor.py
import os
import json
import logging
from telegram.ext import ContextTypes
from config.config import Config
from .connection import ssh_exec

logger = logging.getLogger(__name__)
IP_HISTORY_FILE = "data/ip_history.json"

if not os.path.exists("data"):
    os.makedirs("data")

def get_stored_ips():
    if os.path.exists(IP_HISTORY_FILE):
        try:
            with open(IP_HISTORY_FILE, 'r') as f:
                return json.load(f)
        except:
            pass
    return {"v4": None, "v6": None}

def save_stored_ips(ips):
    try:
        with open(IP_HISTORY_FILE, 'w') as f:
            json.dump(ips, f)
    except Exception as e:
        logger.error(f"Failed to save IP history: {e}")

def get_router_ips():
    v4 = None
    v6 = None
    iface_list = ["wan", "wan_6", "wan6"]
    for iface in iface_list:
        try:
            ubus_cmd = f"ubus call network.interface.{iface} status"
            ubus_res = ssh_exec(ubus_cmd)
            if ubus_res:
                data = json.loads(ubus_res)
                if not v4 and "ipv4-address" in data and len(data["ipv4-address"]) > 0:
                    addr = data["ipv4-address"][0]["address"]
                    if "." in addr:
                        v4 = addr
                if not v6 and "ipv6-address" in data:
                    for addr_obj in data["ipv6-address"]:
                        addr = addr_obj["address"]
                        if ":" in addr and not addr.startswith("fe80"):
                            v6 = addr
                            break
                if not v6 and "ipv6-prefix-assignment" in data:
                    for prefix in data["ipv6-prefix-assignment"]:
                        if "address" in prefix and not prefix["address"].startswith("fe80"):
                            v6 = prefix["address"]
                            break
        except Exception:
            continue
    if not v4:
        v4_res = ssh_exec("/usr/bin/curl -4 -s --max-time 5 icanhazip.com || /usr/bin/curl -4 -s --max-time 5 ifconfig.me")
        if v4_res and v4_res.strip() and " " not in v4_res:
             v4 = v4_res.strip()
    if not v6:
        v6_res = ssh_exec("/usr/bin/curl -6 -s --max-time 5 icanhazip.com || /usr/bin/curl -6 -s --max-time 5 ifconfig.co")
        if v6_res and v6_res.strip() and " " not in v6_res:
             v6 = v6_res.strip()
    if v4 and (len(v4) > 15 or "." not in v4): v4 = None
    if v6 and ":" not in v6: v6 = None
    return v4, v6

async def check_ip_job(context: ContextTypes.DEFAULT_TYPE):
    try:
        current_v4, current_v6 = get_router_ips()
        if not current_v4 and not current_v6:
            return
        stored = get_stored_ips()
        changed = False
        msg = "🚨 **公网 IP 变动通知**\n-------------------\n"
        if current_v4 and current_v4 != stored.get("v4"):
            msg += f"🔴 IPv4: `{current_v4}`\n(旧: {stored.get('v4', '未知')})\n"
            stored["v4"] = current_v4
            changed = True
        if current_v6 and current_v6 != stored.get("v6"):
            msg += f"🔵 IPv6: `{current_v6}`\n(旧: {stored.get('v6', '未知')})\n"
            stored["v6"] = current_v6
            changed = True
        if changed:
            save_stored_ips(stored)
            try:
                await context.bot.send_message(chat_id=Config.ADMIN_ID, text=msg, parse_mode="Markdown")
            except Exception as e:
                logger.error(f"Failed to send IP notification: {e}")
    except Exception as e:
        logger.error(f"Error in check_ip_job: {e}")

def init_plugin(application):
    if application.job_queue:
        application.job_queue.run_repeating(check_ip_job, interval=60, first=10)
        logger.info("IP Monitor Job registered.")

File: pkg/openwrt/ip_monitor.go
package openwrt

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	"github.com/yingxiaomo/homeops/config"
	tele "gopkg.in/telebot.v3"
)

const IPHistoryFile = "data/ip_history.json"

type IPHistory struct {
	V4 string `json:"v4"`
	V6 string `json:"v6"`
}

func GetRouterIPs() (string, string) {
	v4 := ""
	v6 := ""
	ifaces := []string{"wan", "wan_6", "wan6"}

	for _, iface := range ifaces {
		cmd := fmt.Sprintf("ubus call network.interface.%s status", iface)
		res, _ := SSHExec(cmd)
		if res == "" {
			continue
		}

		var data map[string]interface{}
		if err := json.Unmarshal([]byte(res), &data); err == nil {
			if v4 == "" {
				if ipv4Arr, ok := data["ipv4-address"].([]interface{}); ok && len(ipv4Arr) > 0 {
					if addrMap, ok := ipv4Arr[0].(map[string]interface{}); ok {
						if addr, ok := addrMap["address"].(string); ok && strings.Contains(addr, ".") {
							v4 = addr
						}
					}
				}
			}
			if v6 == "" {
				if ipv6Arr, ok := data["ipv6-address"].([]interface{}); ok {
					for _, item := range ipv6Arr {
						if addrMap, ok := item.(map[string]interface{}); ok {
							if addr, ok := addrMap["address"].(string); ok {
								if strings.Contains(addr, ":") && !strings.HasPrefix(addr, "fe80") {
									v6 = addr
									break
								}
							}
						}
					}
				}
				if v6 == "" {
					if ipv6Pre, ok := data["ipv6-prefix-assignment"].([]interface{}); ok {
						for _, item := range ipv6Pre {
							if prefixMap, ok := item.(map[string]interface{}); ok {
								if addr, ok := prefixMap["address"].(string); ok && !strings.HasPrefix(addr, "fe80") {
									v6 = addr
									break
								}
							}
						}
					}
				}
			}
		}
	}

	if v4 == "" {
		res, _ := SSHExec("/usr/bin/curl -4 -s --max-time 5 icanhazip.com || /usr/bin/curl -4 -s --max-time 5 ifconfig.me")
		if res != "" && strings.Contains(res, ".") {
			v4 = strings.TrimSpace(res)
		}
	}
	if v6 == "" {
		res, _ := SSHExec("/usr/bin/curl -6 -s --max-time 5 icanhazip.com || /usr/bin/curl -6 -s --max-time 5 ifconfig.co")
		if res != "" && strings.Contains(res, ":") {
			v6 = strings.TrimSpace(res)
		}
	}

	if len(v4) > 15 || !strings.Contains(v4, ".") {
		v4 = ""
	}
	if !strings.Contains(v6, ":") {
		v6 = ""
	}

	return v4, v6
}

func getStoredIPs() *IPHistory {
	if _, err := os.Stat("data"); os.IsNotExist(err) {
		os.Mkdir("data", 0755)
	}
	data, err := os.ReadFile(IPHistoryFile)
	if err != nil {
		return &IPHistory{}
	}
	var history IPHistory
	json.Unmarshal(data, &history)
	return &history
}

func saveStoredIPs(history *IPHistory) {
	if _, err := os.Stat("data"); os.IsNotExist(err) {
		os.Mkdir("data", 0755)
	}
	data, _ := json.Marshal(history)
	os.WriteFile(IPHistoryFile, data, 0644)
}

func StartIPMonitor(b *tele.Bot) {
	ticker := time.NewTicker(60 * time.Second)
	go func() {
		for range ticker.C {
			checkIPJob(b)
		}
	}()
	log.Println("IP Monitor Job registered.")
}

func checkIPJob(b *tele.Bot) {
	currentV4, currentV6 := GetRouterIPs()
	if currentV4 == "" && currentV6 == "" {
		return
	}

	stored := getStoredIPs()
	changed := false
	msg := "🚨 **公网 IP 变动通知**\n-------------------\n"

	if currentV4 != "" && currentV4 != stored.V4 {
		old := stored.V4
		if old == "" {
			old = "未知"
		}
		msg += fmt.Sprintf("🔴 IPv4: `%s`\n(旧: %s)\n", currentV4, old)
		stored.V4 = currentV4
		changed = true
	}

	if currentV6 != "" && currentV6 != stored.V6 {
		old := stored.V6
		if old == "" {
			old = "未知"
		}
		msg += fmt.Sprintf("🔵 IPv6: `%s`\n(旧: %s)\n", currentV6, old)
		stored.V6 = currentV6
		changed = true
	}

	if changed {
		saveStoredIPs(stored)
		adminID := config.AppConfig.AdminID
		if adminID != 0 {
			menu := &tele.ReplyMarkup{}
			menu.Inline(menu.Row(menu.Data("🔙 返回主菜单", "start_main")))

			_, err := b.Send(&tele.User{ID: adminID}, msg, menu, tele.ModeMarkdown)
			if err != nil {
				log.Printf("Failed to send IP notification: %v", err)
			}
		}
	}
}
